### ISSUE TITLE
How can I deploy Yolov11-obb?

### ISSUE THREAD

**Issue by ZeroLelouch1205** [ultralytics/ultralytics#17900]
### Search before asking

- [X] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

when I user C++ with tensortRT to predict image,the result is different for using python to predict
the python predict code is:
![image](https://github.com/user-attachments/assets/0076aff9-5b07-4e67-adc5-5bb87e6f1e6f)
the C++ predict code is:
preprocess code:
![image](https://github.com/user-attachments/assets/e8fa70c1-6727-4f6c-963a-3280465f300f)
![image](https://github.com/user-attachments/assets/fea62d8d-e1c3-493f-a4e9-5ffefd011313)
after using letterbox to preprocess image, I normalization the image using this code:
![image](https://github.com/user-attachments/assets/2fbfc6c2-f854-41fa-828f-82a387665241)
and then invert image to tensor for model input：
![image](https://github.com/user-attachments/assets/ddbab6d3-3097-4d8a-9103-b551f8b36d9b)
Which step did I do wrong caused the C++ inference results to be inconsistent with Python, and the confidence gap was too large？the python inference result is ok,and it's confidence is 0.97; but i use c++ code to inference, the result is ok、confidence is 0.81

**Comment by UltralyticsAssistant** [comment]
👋 Hello @ZeroLelouch1205, thank you for your interest in Ultralytics 🚀!

We recommend visiting the [Docs](https://docs.ultralytics.com) for detailed resources and examples, including guides for using YOLO with different frameworks. For questions about custom deployments like yours using C++ and TensorRT, please ensure you’ve followed the [Inference Guide](https://docs.ultralytics.com/tasks/detect/#inference) for Python and are accurately replicating input preprocessing steps in your C++ code.

If this is a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum_reproducible_example/) to help us debug effectively. This should ideally include your model configurations, test images, and full reproducible code snippets for both Python and C++ pipelines.

If this is a custom deployment ❓ Question, please ensure you provide:
1. Detailed descriptions of your preprocessing methods (e.g., resizing, normalization, and tensor conversion).
2. Inference logs and outputs from both Python and C++ to help identify discrepancies.
3. Verification that the correct model weights are being used in both environments.

In the meantime, here are some steps to try:
- Upgrade to the latest `ultralytics` version and ensure you meet all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) to confirm this issue isn’t resolved in the latest release:

```bash
pip install -U ultralytics
```

- Compare the letterbox resizing and normalization logic between Python and C++ implementations. These are common areas for subtle mismatches resulting in confidence differences.

💡 For real-time assistance, consider joining our [Discord](https://discord.com/invite/ultralytics) community 🎧. You may also find helpful insights on [Discourse](https://community.ultralytics.com) or the Ultralytics [Subreddit](https://reddit.com/r/Ultralytics) for similar use cases.

## Verified Environments

YOLO runs reliably in these environments, which you may want to consider for cross-validation:
- **Notebooks**: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Pre-built Cloud Instances**: [GCP](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/) or [AWS](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/).
- **Docker**: Try using our preconfigured [Docker Image](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## CI Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yaml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yaml/badge.svg" alt="Ultralytics CI"></a>

Stay tuned! 🚀 An Ultralytics engineer will review your query and provide further assistance soon.

**Comment by Y-T-G** [comment]
Are you comparing with TensorRT in Ultralytics?

**Comment by ZeroLelouch1205** [comment]
> Are you comparing with TensorRT in Ultralytics?

yes, i want to use tensorRT to deploy the model, but the result is not same as python inference;I debug the python predict,find the image preprocess by letterbox, so i do the same thing in C++ code
![image](https://github.com/user-attachments/assets/cd663bfa-5323-421a-bf8e-8224614e7a41)
![image](https://github.com/user-attachments/assets/da8efedf-f041-4487-9a1f-34b99dd44591)

**Comment by ZeroLelouch1205** [comment]
> @ZeroLelouch1205 the discrepancy between Python and TensorRT inferences often results from differences in preprocessing or data formatting. Ensure your C++ preprocessing replicates Python’s exactly, including mean/standardization values, resizing strategy, and data layout (e.g., RGB vs. BGR channels). For additional debugging, export to ONNX (`model.export(format="onnx")`) and verify intermediate outputs match both environments before converting to TensorRT. More details can be found in our [NVIDIA Jetson guide](https://docs.ultralytics.com/guides/nvidia-jetson/).

i find the python code do not mean/standardization values and resizing strategy is letterbox?

**Comment by ZeroLelouch1205** [comment]
> @ZeroLelouch1205 the discrepancy between Python and TensorRT inferences often results from differences in preprocessing or data formatting. Ensure your C++ preprocessing replicates Python’s exactly, including mean/standardization values, resizing strategy, and data layout (e.g., RGB vs. BGR channels). For additional debugging, export to ONNX (`model.export(format="onnx")`) and verify intermediate outputs match both environments before converting to TensorRT. More details can be found in our [NVIDIA Jetson guide](https://docs.ultralytics.com/guides/nvidia-jetson/).

i find the python code do not mean/standardization values and resizing strategy is letterbox, and i use onnx model path to initial Yolo class , i think it is not convert to tensorRT cause the different result;
so i want to know the preprocessing in python,could u tell me?

**Comment by ZeroLelouch1205** [comment]
> > the python inference result is ok,and it's confidence is 0.97; but i use c++ code to inference, the result is ok、confidence is 0.81
> 
> Is this the max confidence you get?

yes,the classify task or obb task have the same question, the obb task also have wrong conf and have wrong rect but same conf

**Comment by glenn-jocher** [comment]
It seems like the issue lies in the preprocessing or decoding steps in your C++ pipeline. Ensure that the letterbox resizing, normalization values (if applicable), and data ordering (e.g., RGB vs. BGR) precisely replicate the Python implementation. You can compare intermediate outputs (e.g., preprocessed images and model outputs) between Python and C++ to identify discrepancies. If you haven't yet, try exporting the model to ONNX and verifying inference outputs with Python before converting it to TensorRT for your C++ implementation. Let us know if further clarification is needed!

**Comment by ZeroLelouch1205** [comment]
> It seems like the issue lies in the preprocessing or decoding steps in your C++ pipeline. Ensure that the letterbox resizing, normalization values (if applicable), and data ordering (e.g., RGB vs. BGR) precisely replicate the Python implementation. You can compare intermediate outputs (e.g., preprocessed images and model outputs) between Python and C++ to identify discrepancies. If you haven't yet, try exporting the model to ONNX and verifying inference outputs with Python before converting it to TensorRT for your C++ implementation. Let us know if further clarification is needed!

i debug the python pipeline,the letterbox resizing result is same as c++ letterbox resizing;
and the c++ mat2tensor pipline is this:
![image](https://github.com/user-attachments/assets/c0e9eb44-7c41-46ec-acfe-3c7add360ff5)
![image](https://github.com/user-attachments/assets/56966b62-c408-4e5a-bb18-4d5762304496)
![image](https://github.com/user-attachments/assets/162376c2-eb2d-4806-be80-90257105f670)
the python pipline is:
![image](https://github.com/user-attachments/assets/c9657be7-14ce-4f6a-8909-5e1a3b2bb82b)

**Comment by glenn-jocher** [comment]
Based on your debugging, since the letterbox resizing and tensor conversion steps appear consistent between Python and C++, the confidence discrepancy may be rooted in subtle numerical differences introduced during operations like normalization or datatype handling. To isolate the issue, validate the intermediate values (e.g., normalized pixel values and final input tensor to the model) between Python and C++ precisely. Also, confirm that the model's floating-point precision (FP32 vs. FP16) and the activation/output processing (e.g., sigmoid/scaling for confidence scores) are identical between both pipelines.

**Comment by djdll** [comment]
I deployed fp16 with ncnn on cpu, but I found that 32-bit speed is very slow, 64-bit speed is much faster, do you know why? How can fast detection be achieved in 32 bits

**Comment by glenn-jocher** [comment]
The slower performance on 32-bit may stem from the lack of hardware optimizations or computational overhead compared to 64-bit systems. For faster detection on 32-bit, ensure your NCNN build is optimized for your specific hardware (e.g., enabling SIMD instructions). Additionally, consider using INT8 quantization if supported, as it reduces computation and improves speed.
